File: ClientApp/src/app/models/song.ts
import { Chord } from './chord';

export interface ChordSequenceItem {
  chordId: number;
  duration: number; // beats duration
  fret: number;     // starting fret position
}

export interface TablatureNote {
  beat: number;     // position in beats
  string: number;   // 0-5 (E, B, G, D, A, E)
  fret: number;     // fret number
  duration: number; // beats duration
}

export interface MusicalElement {
  type: 'chord' | 'note' | 'tab';
  beat: number;      // position in beats
  duration: number;  // beats duration
  
  // For chords
  chordId?: number;
  fret?: number;     // starting fret position
  
  // For single notes
  string?: number;   // 0-5 (E, B, G, D, A, E)
  noteFret?: number; // fret number for the note
  
  // For tablature (all strings at once)
  tabPositions?: string; // JSON string like "[0,0,0,0,7,0]"
}

export interface Song {
  id?: number;
  name: string;
  artist?: string;
  tuning?: string;
  tempo?: number;
  sequenceJson?: string;
  tablatureJson?: string;
  chords?: Chord[];
}

File: Models/Song.cs
namespace GuitarGuide.Models;

public class Song
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Artist { get; set; }
    public string? Tuning { get; set; }
    public int? Tempo { get; set; }
    
    // JSON array with chord sequence: [{ chordId: 1, duration: 4, fret: 0 }, ...]
    public string? SequenceJson { get; set; }
    
    // JSON array with tablature notes: [{ beat: 0, string: 0, fret: 3, duration: 1 }, ...]
    public string? TablatureJson { get; set; }
    
    public ICollection<SongChord> SongChords { get; set; } = new List<SongChord>();
}

// Junction table for many-to-many relationship
public class SongChord
{
    public int SongId { get; set; }
    public Song Song { get; set; } = null!;
    
    public int ChordId { get; set; }
    public Chord Chord { get; set; } = null!;
}
